File: coding_int/chapter_1/1_6_StringCompression.py
from collections import Counter

def string_compression(stringa):

    pos = 0
    count = 0

    dict_test = Counter(stringa)
    if (len(dict_test.keys()) == len(stringa)):
        return stringa

    result = ""
    for item in stringa:
        if pos == 0:
            holder = item
            count = 1
            result += item

        else:
            if item == holder:
                count += 1
            else:
                result += str(count)
                holder = item
                result += item
                count = 1

            if (pos == (len(stringa) - 1)):
                result += str(count)

        pos += 1
    
    return result

if __name__ == "__main__":

    print(string_compression("aabccaabbcc"))
    print(string_compression("baltic"))
    print(string_compression("rabbit"))
    print(string_compression("eat eat eat"))
    print(string_compression("aabccaabbcc"))
    print(string_compression("bacon toot"))

File: coding_int/chapter_1/1_6_StringCompression.cpp
#include<iostream>
#include<string>

std::string compressString(std::string toTest) {

    char holder;
    int counter = 0;
    bool shortened = false;
    std::string answer_string = "";


    for(std::string::size_type i = 0; i < toTest.length(); ++i ) {

        if (i == 0) {
            holder = toTest[0];
            counter = 1;
            answer_string.append(1, holder);
        } else {
            if (holder == toTest[i]) {
                ++counter;
                shortened = true;
            } else {
                answer_string.append(std::to_string(counter));
                counter = 1;
                holder = toTest[i];
                answer_string.append(1, holder);
            }
        }

        if (i == toTest.length() - 1) {
            answer_string.append(std::to_string(counter));
        }


    }
    if (shortened) {
        return answer_string;
    }
    return toTest;

}

int main() {

    std::string a = "aabccaabbcc";
    std::string b = "baltic";
    std::string c = "rabbit";
    std::string d = "eat eat eat";
    std::string e = "aabccaabbcc";
    std::string f = "bacon toot";

    std::cout << compressString(a) << std::endl;
    std::cout << compressString(b) << std::endl;
    std::cout << compressString(c) << std::endl;
    std::cout << compressString(d) << std::endl;
    std::cout << compressString(e) << std::endl;
    std::cout << compressString(f) << std::endl;


}